feat(openapi): support OpenAPI version 3.1

SF97 opened the pull request:
- This PR adds support for OpenAPI version 3.1. This has been tried before, as described in https://github.com/cdimascio/express-openapi-validator/issues/573, but these attempts failed. Mainly due to a bug in AJV that makes it incompatible with OpenAPI version 3.1 - https://github.com/ajv-validator/ajv/issues/1745

This PR circumvents AJV issue by replacing all $dynamicAnchor keywords in OpenAPI schema with a direct reference, using $ref. This has been suggested by @Jackman3005 in https://github.com/cdimascio/express-openapi-validator/issues/573#issuecomment-1869810476, which originally came from @seriousme, which already made this workaround in his own OpenAPI schema validation library, https://github.com/seriousme/openapi-schema-validator. From what I understand this shouldn't be an issue. Using $dynamicAnchor would allow this lib to replace OpenAPI specification definitions, which it does not do.

# Changes
- `openapi.schema.validator.ts` now creates a different AJV instance that supports JSON schema 2020-12 when it detects version 3.1. This AJV instance accepts the format `media-range` but does not validate its content. This validation can be made in another PR
- Adds OpenAPI specification 3.1 with $dynamicAnchor references changed to $ref
- Change test scripts to use Mocha's `--extension` flag. With the previous setup, the new tests weren't being picked up by the runner

# OpenAPI 3.1 support

- [x] Support webhook property without any defined paths
- [x] Support an API with only components
- [x] Fully complete Open API 3.1 schema in type.ts
- [x] Support "summary" in Info Object
- [x] Support "identifier" field for SPDX licenses in License Object
- [x] Ensure "nullable" property does not work anymore. Null type works instead
- [x] Ensure Request Body validation works in GET HTTP methods
- [x] Ensure "default" must exist in server variable "enum" values
- [x] Ensure it is possible to create an Operation Object without `responses`
- [x] Support "pathItems" in Components Object

There are some more changes introduced by Open API 3.1, but they're schema changes that are validated by AJV, so tests, although important, aren't as critical as the ones above. They are:

- Ensure mutual TLS works correctly
- Ensure extension prefixed with "x-oas-" is not allowed
- Ensure "exclusiveMaximum" and "exclusiveMinimum" do not accept boolean values
- Ensure server variable "enum" can not be empty
- Ensure "jsonSchemaDialect" is supported
- Ensure "style", "explode" and "allowReserved" for multipart/form-data are allowed as media types


Closes #573 #755

SF97 committed:
feat(openapi): support version 3.1

cdimascio commented:
Thank you! can you help build the scaffolding and establish a pattern for 3.1 tests, enabling myself and the community to contribute and expand on them

SF97 commented:
> Thank you! can you help build the scaffolding and establish a pattern for 3.1 tests, enabling myself and the community to contribute and expand on them

Yes, I can do that! I've been really busy lately so I haven't got around to that yet

I think I'll have to time to pick it this weekend

SF97 committed:
test(openapi_3.1): ensure that an API with webhooks and no routes is supported

SF97 committed:
feat(openapi_3.1): adds open api 3.1 type

SF97 committed:
chore(test-scripts): run mocha with --extension instead of glob to pick up subdirectories

Mocha was not picking up the tests in subdirectories with the provided glob. Adding --extension with
the tests extension and setting the root test folder in tests fixed it

SF97 commented:
Hello!

I just pushed the first test for OpenAPI 3.1, where we test for webhook support without any API routes, in which I found some bugs immediately :rofl: but at least I was able to fix it and the tests are passing :partying_face: 

I created the folder `test/openapi_3.1` for version 3.1 related tests. I've also added the interface `DocumentV3_1` for a OpenAPI 3.1 compatible type, and refactored all references to the OpenAPI document to union both 3.0 and 3.1. I found this pretty ugly, maybe we should create an abstraction around it. Anyway, what I wanted was to at least do the first tests so we can all contribute to this PR :) 

I'll try to contribute as much as I can in the upcoming days. Anything that I can help with, just let me know

cdimascio commented on src/middlewares/parsers/schema.preprocessor.ts:
```diff
@@ -108,22 +108,28 @@ export class SchemaPreprocessor {
-    const r = this.gatherSchemaNodesFromPaths();
+    let r;
+
+    if (this.apiDoc.paths) {
```
can you help me understand this. does `paths` not exist in 3_1

SF97 replied:
Hello

It does exist in 3_1_, but it's optional. Since OpenAPi version `3.1.0-rc0` "An OpenAPI Document now requires at least one of paths, components or webhooks to exist at the top level. While previous versions required paths, now a valid OpenAPI Document can describe only webhooks, or even only reusable components. Thus, an OpenAPI Document no longer necessarily describes an API." [1]

[1] https://github.com/OAI/OpenAPI-Specification/releases/tag/3.1.0-rc0

cdimascio reviewed (commented):
this looks good. one question regarding apiDocs.paths i.e. why might it be missing and what is the behavior when it is

SF97 committed:
test(openapi-3.1): adds test to ensure an API with only components is considered valid

SF97 committed:
test(openapi-3.1): remove unnecessary import

SF97 committed:
test(openapi-3.1): add support for summary in info object

SF97 committed:
test(openapi-3.1): add support for identifier in license

SF97 commented:
Hello

Just added the list of things we need to ensure that work in OpenAPI 3.1. I based it on OpenAPI's own changelog. I'm currently adding tests to ensure that everything works with the 3.1 spec just fine. Just added some and I'm happy to receive contributions :)

SF97 committed:
test(openapi_3.1): ensure API with type set to null works correctly

SF97 committed:
test(open_api3.1): ensure that methods with non-explicit semantics allow request body

SF97 committed:
test(open_api3.1): ensure 500 is returned when server variable has no default

SF97 committed:
feat(openapi_3.1): ensure API supports an endpoint without response

SF97 committed:
Merge remote-tracking branch 'upstream/master' into feat/openapi-3.1-support

SF97 committed:
feat(openapi_3.1): add full type support for open api 3.1

SF97 committed:
test(openapi_3.1): adds test for path item support in components

SF97 commented:
Hello @cdimascio 

I've added the most important tests. There are some more things to validate, but they're assured by AJV, thus I don't think they're critical to have this feature.

But I need help in one thing before shipping. I've added a test to ensure that `pathItems` is supported in components object, but it's failing. The YAML (`test/openapi_3.1/resources/components_path_items.yaml`) is fine, since it's parsed correctly in https://editor-next.swagger.io/, and it's according to the spec, but AJV is saying it's invalid, and I'm not sure why. It throws this error:
![image](https://github.com/cdimascio/express-openapi-validator/assets/10375710/61935a2f-56ea-4903-8b9e-65fa873de4fa)

Can you give a little help here to ensure that we'll support this feature? After this test is passing we can proceed with a full review and add support for OpenAPI 3.1 :partying_face: 

Thank you

SF97 commented:
@medranocalvo

Thank you very much! I took a look at the schema, and it looks like it was a bug at the schema level, which has been fixed in https://github.com/OAI/OpenAPI-Specification/pull/3355, i.e. the YAML was fine, but the schema wasn't validating it correctly. I've updated the schema to the newest, and it no longer throws a schema error :sunglasses:.

The test is still failing, but I suspect what it is... I'm going to work on it

SF97 committed:
fix(openapi_3.1_schema): update schema to fix bug

SF97 committed:
feat(openapi_3.1): support reusable path items

SF97 commented:
And it is fixed :) All tests are passing

That means this PR is complete AFAIK and can be reviewed and merged to fully support Open API 3.1, including reusable path items which I just fixed

@cdimascio could you please review and let me know if there's anything that needs to be changed?

SF97 committed:
style(linting): fix linting issues

SF97 commented on src/middlewares/openapi.request.validator.ts:
```diff
@@ -21,6 +21,7 @@ import {
 import { BodySchemaParser } from './parsers/body.parse';
 import { ParametersSchemaParser } from './parsers/schema.parse';
 import { RequestParameterMutator } from './parsers/req.parameter.mutator';
+import { debug } from 'console';
```
fixed

SF97 committed:
style(openapi): improve readability of version validation

SF97 commented on src/framework/openapi.schema.validator.ts:
```diff
@@ -29,18 +32,31 @@ export class OpenAPISchemaValidator {
-    addFormats(v, ['email', 'regex', 'uri', 'uri-reference']);
-
-    const ver = opts.version && parseInt(String(opts.version), 10);
+    const ver = opts.version && parseFloat(String(opts.version));
```
Thanks for the suggestions, that's read a lot better :) I've changed the code and tweaked a little bit for readability with some of your suggestions

SF97 committed:
docs(schema-validator): clearly state why media-range attribute is not defined

SF97 commented on src/framework/openapi.schema.validator.ts:
```diff
@@ -29,18 +32,31 @@ export class OpenAPISchemaValidator {
+    } else if (ver === 3.1) {
+      schema = openapi31Schema;
+      ajvInstance = new Ajv2020(options);
+      ajvInstance.addFormat('media-range', true); // TODO: Validate media-range format as defined in https://www.rfc-editor.org/rfc/rfc9110.html#name-collected-abnf
```
Added an explanation on why it's not validated

cdimascio commented:
Thank you

SF97 committed:
Merge branch 'master' into feat/openapi-3.1-support

Jackman3005 commented:
Hi @cdimascio @mdmower @SF97 I'm still looking forward to testing this as I think it will help with a number of issues I'm facing around `nullable`.

Where are things at right now? Is there a list of outstanding tasks we could make or is it ready for a beta release?

Thanks in advance 🙏

SF97 commented:
> Hi @cdimascio @mdmower @SF97 I'm still looking forward to testing this as I think it will help with a number of issues I'm facing around `nullable`.
> 
> Where are things at right now? Is there a list of outstanding tasks we could make or [is it ready for a alpha release](https://github.com/cdimascio/express-openapi-validator/issues/573#issuecomment-1874742468)?
> 
> Thanks in advance 🙏

Hello

From my end, it's ready. I've added all the test cases and fixed the bugs that I've found along the way. If we can't do an alpha release, you can always install it from this branch to test it out locally

huilingwei commented:
@cdimascio @mdmower @SF97 Hi, is there any plan when this feature will be released?

patricktyndall commented:
chiming in to express interest, particularly in `unevaluatedProperties`. Thanks all for the work

cdimascio commented:
The content in this PR is now available on branch `oas-3.1` and package `express-openapi-validator@6.0.0-alpha.1`. please try it out.

```
# experimental OAS 3.1 in alpha (contributions welcome - see branch `oas-3.1` and pr-882)
npm install express-openapi-validator@6.0.0-alpha.1
```

To contribute submit a PR against `oas-3.1`

Eileen-34 commented:
Do you know if koa openapi validator, which is provided by Express, support 3.1 v OpenAPI Spec?

https://github.com/cdimascio/express-openapi-validator/tree/lerna-fastify/packages/koa-openapi-validator

sergio-ferreira-jscrambler commented:
Hey @cdimascio 

I think this is ready for a release, do you agree? We've given enough time for feedback and bug fixing, if needed

SF97 commented:
Hey @cdimascio

I think this is ready for a release, do you agree? We've given enough time for feedback and bug fixing, if needed

uwinkelvos commented on src/framework/types.ts:
```diff
@@ -186,6 +186,19 @@ export namespace OpenAPIV3 {
+    pathItems?: { [path: string]: PathItemObject | ReferenceObject }
+  }
+
+  export interface DocumentV3_1 extends Omit<DocumentV3, 'paths' | 'info' | 'components'> {
```
```suggestion
  export interface DocumentV3_1 extends Omit<DocumentV3, 'paths' | 'info' | 'components', 'openapi'> {
      openapi: `3.1.${p}`;
```
To make this also compile type safe:
could then also be applied above in 3.0 with `openapi: `3.0.${p}`;`

cdimascio commented:
merged via #1027
